### PR TITLE
WIP: oneDAL error handling impl

### PR DESCRIPTION
**Pull request by mchernov-intel**
moved from: https://github.com/rlnx/daal/pull/6

### CODE REVIEW

File: cpp/oneapi/dal/backend/interop/error_convertor.hpp [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+
+
+#include "oneapi/dal/exception.hpp"
+#include "daal/include/services/error_handling.h"
+#include "daal/include/services/internal/status_to_error_id.h"
+
+#include <iostream>
```
Comment (mchernov-intel) [inline cpp/oneapi/dal/backend/interop/error_convertor.hpp:22]: should be removed

File: cpp/daal/include/services/internal/status_to_error_id.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+*******************************************************************************/
+
+#ifndef __SERVICES_INTERNAL_STATUS_TO_ERROR_ID_H__
+#define __SERVICES_INTERNAL_STATUS_TO_ERROR_ID_H__
+
+namespace daal
+{
+namespace services
+{
+namespace interface1
```
Comment (michael-smirnov) [inline cpp/daal/include/services/internal/status_to_error_id.h:25]: you missed `internal` namespace

File: cpp/daal/src/services/error_handling.cpp [hunk 3]
```diff
@@ -621,6 +621,19 @@ services::ErrorCollectionPtr Status::getCollection() const
     return services::ErrorCollectionPtr(new services::ErrorCollection());
 }
 
+namespace interface1
+{
+ErrorID get_error_id(const Status & s)
```
Comment (michael-smirnov) [inline cpp/daal/src/services/error_handling.cpp:626]: Do we really need to put this function under `interfaceX` namespace?

File: cpp/oneapi/dal/algo/pca/backend/cpu/train_kernel_cov.cpp [hunk 4]
```diff
@@ -79,6 +80,10 @@ static train_result call_daal_kernel(const context_cpu& ctx,
         *daal_means,
         *daal_variances);
 
+    if (!status) {
+        interop::status_to_exception(status, [](const daal::services::Status& s) { });
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/algo/pca/backend/cpu/train_kernel_cov.cpp:84]: You can provide an overload of `status_to_exception()` which don't require user to provide empty lambda.
I suggest to name it differently, e.g. `interop::status_to_exception_default()`, so user will be aware that he calls only default conversion, while in common case he has to write his own conversion.

Comment (mchernov-intel) [reply cpp/oneapi/dal/algo/pca/backend/cpu/train_kernel_cov.cpp:84]: @rlnx

**Comment by mchernov-intel** [comment]
@michael-smirnov please, review

File: cpp/oneapi/dal/backend/interop/error_convertor.hpp [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+        case ErrorID::ErrorPrecomputedMaxNotAvailable:
+            throw unavailable_error(description);
+        case ErrorID::ErrorSourceDataNotAvailable:
+            throw unavailable_error(description);
+        case ErrorID::ErrorFeatureNamesNotAvailable:
+            throw unavailable_error(description);
+        case ErrorID::ErrorAccessUSMPointerOnOtherDevice:
+            throw unavailable_error(description);
+
+        ////////////////////////////////////////////
```
Comment (mchernov-intel) [inline cpp/oneapi/dal/backend/interop/error_convertor.hpp:351]: should be removed

File: cpp/oneapi/dal/backend/interop/error_convertor.hpp [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+namespace oneapi::dal::backend::interop {
+
+void status_to_exception_default(const daal::services::Status &s) {
+    using namespace daal::services;
+    using namespace daal::services::internal;
+
+    const ErrorID error = get_error_id(s);
+    const char * description = s.getDescription();
+
+    switch (error) {
```
Comment (PetrovKP) [inline cpp/oneapi/dal/backend/interop/error_convertor.hpp:31]: Maybe:
```c++
const char * description = s.getDescription();
if (!error ) {
  throw invalid_argument(description);
}
```

Comment (mchernov-intel) [reply cpp/oneapi/dal/backend/interop/error_convertor.hpp:31]: There are different types of error classes.

**Pull request closed without merging**